src/include/storage/proc.h

```diff
@@ -186,6 +186,7 @@ struct PGPROC
 	/* Lock manager data, recording fast-path locks taken by this backend. */
 	uint64		fpLockBits;		/* lock modes held for each fast-path slot */
 	Oid			fpRelId[FP_LOCK_SLOTS_PER_BACKEND];		/* slots for rel oids */
+	bool		fpHoldTillEndXact[FP_LOCK_SLOTS_PER_BACKEND];		/* slots for gdd flags */
```

Comment:
bool is implemented as char. I think char is OK? @pivotal-ning-yu

Reply:
Using bool or char will add 16 bytes to PGPROC struct, that's too much. Why is `fpLockBits` defined as `uint64` instead of `uint8 [16]`? The same purpose.

Reply:
@pivotal-ning-yu 
You mean using a integer's bit to mimic array?

It is tricky because it depends on FP_LOCK_SLOTS_PER_BACKEND.

Reply:
fast path implementation is already playing this trick as `fpLockBits`. Wasting memory in `PGPROC` struct should always be avoided.

Reply:
`FP_LOCK_SLOTS_PER_BACKEND` is not enough for your purpose with `fpHoldTillEndXact`.  Note that one slot here can stand for at most `FAST_PATH_BITS_PER_SLOT` locks.

Reply:
Yes, that is what `FAST_PATH_BITS_PER_SLOT` is used for. Is the `holdTillEndXact` info per lockmode or per lock?